### ISSUE TITLE
Solving "Please import the top-level fullcalendar lib before attempting to import a plugin."

### ISSUE THREAD

**Issue by arshaw** [fullcalendar/fullcalendar#6371]
This JS error occurs when a FullCalendar plugin, like `@fullcalenar/daygrid` or `@fullcalendar/timeline`, is imported before `@fullcalendar/core` or one of the connector libraries, like React, Vue, or Angular. I will backlink tickets related to each of these three frameworks to this main ticket.

It is obviously not good to have ordering restrictions for an end-developer's `import` statements. The goal is to do away with this restriction.

The restriction exists for reasons that are internal to FullCalendar's rendering code: FullCalendar is designed to work with an abstract component/jsx implementation. Either React or Preact. When FullCalendar's component classes are being created at the top-level runtime, they need to know what `Component` class to subclass. Thus, either React or Preact needs to be available at that time.

Version 6 will sidestep this issue by using (p)react hooks for rendering. With hooks, there is no longer the need to subclass a Component class.

**Comment by AdrianMrn** [comment]
For anyone else using [Vite](https://vitejs.dev/), this issue breaks the normal `vite` dev build process because of how dependencies are imported in Vite dev mode. I don't know what exactly is causing the problem, so somebody who knows the configuration options better than I do might be able to solve this issue.

You can temporarily work around this by using `yarn vite build -w`, which creates a production build but still watches the files.

**Comment by duckbrain** [comment]
@AdrianMrn See https://github.com/fullcalendar/fullcalendar-vue/issues/152#issue-918047023 for Vite workaround.

**Comment by AdrianMrn** [comment]
@duckbrain Thanks for pointing me in the right direction! I'm bundling a React application with Vite, so this fix didn't work for me. However, adding this line does fix it: `import '@fullcalendar/react/dist/vdom';` 🎉

**Comment by braindeaf** [comment]
I am experiencing this error but I am somewhat baffled since this only started complaining when I added moment-timezone and the @fullcalendar/react is imported first unless I am missing something.

<img width="736" alt="Screenshot 2022-07-01 at 10 50 26" src="https://user-images.githubusercontent.com/10952/176871479-1c20bdb8-5f60-4d45-bee4-dce5969a66cd.png">


```
import React from 'react';
import FullCalendar from '@fullcalendar/react';
import dayGridPlugin from '@fullcalendar/daygrid';
import listPlugin from '@fullcalendar/list';
import timeGridPlugin from '@fullcalendar/timegrid';
import interactionPlugin from '@fullcalendar/interaction';
import momentTimezonePlugin from '@fullcalendar/moment-timezone';
```

**Comment by casually-creative** [comment]
I am receiving the same error when using snowpack to compile my files. I assume the underlying issue is the same as the one with Vite. When adding `import '@fullcalendar/react/dist/vdom';` as [AdrianMrn](https://github.com/AdrianMrn) suggested, the issue was fixed.

**Comment by Srounsroun** [comment]
Thanks for the answer, it works well on development mode.

I have one issue now, when building for production using vite.

```
Error: Minified React error #185; visit https://reactjs.org/docs/error-decoder.html?invariant=185 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Jg (react-dom.production.min.js:237:48)
    at Object.enqueueSetState (react-dom.production.min.js:133:211)
    at C$4.setState (react.production.min.js:12:369)
    at Ar.handleSlatCoords [as onCoords] (main.js:999:19)
    at $r.updateSizing (main.js:449:23)
    at $r.componentDidUpdate (main.js:435:14)
    at Yi$1 (react-dom.production.min.js:219:502)
    ....
```

Seems that the issue is in this methode: 

```
TimeColsSlats.prototype.componentDidUpdate = function () {
        this.updateSizing();
    };
    TimeColsSlats.prototype.componentWillUnmount = function () {
        if (this.props.onCoords) {
            this.props.onCoords(null);
        }
    };
```

Do someone already had this problem ? maybe someone have some clues on this

**Comment by Srounsroun** [comment]
It seems that using initialView="timeGridWeek" with fullcalendar react component and vitejs is not working on production build..
I didn't found something yet to fix it :(

**Comment by Fraganya** [comment]
> I am receiving the same error when using snowpack to compile my files. I assume the underlying issue is the same as the one with Vite. When adding `import '@fullcalendar/react/dist/vdom';` as [AdrianMrn](https://github.com/AdrianMrn) suggested, the issue was fixed.

This solved my issue. Dont really understand how it works Thought. Thanks

**Comment by arshaw** [comment]
Released in [v6.0.0](https://github.com/fullcalendar/fullcalendar/releases/tag/v6.0.0)

**Comment by nwabekeyi** [comment]
import '@fullcalendar/react/dist/vdom

This fixed it.
